### PR TITLE
fix: Fix issue 25285 max insufficient funds for gas

### PR DESCRIPTION
**Pull request by ejwessel**
## **Description**

When performing a send or swap+send there are two outstanding issues:
1. swap+send, nativeToken -> ERC20 and "insufficient funds for gas" error message appears. This behavior is incorrect estimation and should only occur when there are insufficient funds for simple send. In other words, the current logic is treating nativeToken -> ERC20 swaps as if they were nativeToken -> nativeToken swaps, which is causing this error.
1. when swapping between src assets the amount mode max is maintained. Amounts should reset between differently chosen srcAssets.

Fixes are:
* Hiding the max button when the source token is native, but the destination token is an ERC20
* Resetting the srcAmount whenever the srcAsset changes

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25574?quickstart=1)

## **Related issues**

Fixes: 
* https://github.com/MetaMask/metamask-extension/issues/25285
* https://github.com/MetaMask/metamask-extension/issues/25594

## **Manual testing steps**

For each every combination of srcToken to destToken either being nativeToken or an ERC20
Verify that the insufficient funds message does not show unless it is nativeToken -> nativeToken and there is insufficient funds for gas

when maxing srcToken (current behavior)
| srcToken | destToken | 
| --- | --- | 
| nativeToken | nativeToken |
| nativeToken | ERC20 |  
| ERC20 | nativeToken | 
| ERC20 | ERC20 |

1.  click on 'Send'
3. perform a 'Send' or 'Swap + Send' based on the above table
4. verify that the error message appears only in the situation where there is insufficient native token to pay for gas for a 'Send'

### **Before**

Screenshot 2024-06-20 at 00 01 00

### **After**

Screenshot 2024-06-28 at 15 15 43
Screenshot 2024-06-28 at 15 15 46
Screenshot 2024-06-28 at 15 15 53
Screenshot 2024-06-28 at 15 15 56
Screenshot 2024-06-28 at 15 16 16
Screenshot 2024-06-28 at 15 16 24
Screenshot 2024-06-28 at 15 16 32

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by sleepytanya** [comment]
Works great:

https://github.com/MetaMask/metamask-extension/assets/104780023/cd61171c-3e94-45a1-b9d1-991bd9f4b728

What are these 'not connected' circles next to the addresses on the Send page popup? One of the accounts is connected to the the dapp on selected network.

Screenshot 2024-06-27 at 19 14 49
@ejwessel

**Comment by ejwessel** [comment]
> Works great:
> 
>  max.mov 
> What are these 'not connected' circles next to the addresses on the Send page popup? One of the accounts is connected to the the dapp on selected network.
> 
> Screenshot 2024-06-27 at 19 14 49 @ejwessel

I'm not sure. I have a bunch of accounts that I use for testing and these are probably them. This is unrelated to this fix though.

**Review by sleepytanya — DISMISSED** [review]

**Comment by sleepytanya** [comment]
'Max' amount successfully resets to 0 when another source token is selected:

https://github.com/MetaMask/metamask-extension/assets/104780023/09130e58-a55e-4cfc-8cdd-9b54a2ae16da

**Comment by ejwessel** [comment]
* sigh * apparently the error still shows if you wait when maxing native -> native.

**Review by jpuri — DISMISSED** [review]
Changes look good 👍

**Comment by sleepytanya** [comment]
If gas is updated while user is on the send+swap page the amount is not recalculated and 'insufficient funds' are shown.

Screenshot 2024-07-02 at 12 27 37

https://github.com/MetaMask/metamask-extension/assets/104780023/cfbb042b-2018-47ca-9d8c-cdd42c0f4f5e

**Comment by sleepytanya** [comment]
'Max' amount option works as expected in all the options from the table below:

Screenshot 2024-07-04 at 00 16 58

 'Max' option is not displayed for the ETH -> ERC20 send

Screenshot 2024-07-04 at 00 15 28



 For the native->native send amount is successfully updated with the gas estimation changes:


https://github.com/MetaMask/metamask-extension/assets/104780023/f921d126-9716-4503-bbc1-4be890e64932

File: ui/components/multichain/asset-picker-amount/asset-picker-amount.tsx [hunk 1]
```diff
@@ -135,6 +137,10 @@ export const AssetPickerAmount = ({
     borderColor = BorderColor.primaryDefault;
   }
 
+  const shouldHideMaxClearButton =
```
Comment (BZahory) [inline ui/components/multichain/asset-picker-amount/asset-picker-amount.tsx:140]: Can we rename this to `isSwapAndSendFromNative` or something similar for clarity?

Comment (ejwessel) [reply ui/components/multichain/asset-picker-amount/asset-picker-amount.tsx:140]: I believe the current name of the flag effectively indicates its purpose, which is to hide the max button only under the following condition. Though you are indicating it's not clear, can you help me understand what specifically is unclear about the current name?

**Review by micaelae — DISMISSED** [review]

File: ui/components/multichain/asset-picker-amount/asset-picker-amount.tsx (hunk 1 above)
Comment (BZahory) [reply ui/components/multichain/asset-picker-amount/asset-picker-amount.tsx:140]: I think it's more clear to name boolean variables based on tracked behavior vs the resultant action. Naming this `shouldHideMaxClearButton` doesn't account for a) other conditions hiding the max clear button (i.e. `shouldHideMaxClearButton` could be `false`, but we would hide it anyways if there isn't an amount handler available) and b) makes it less clear that we can use this variable for other implementations in the future.

**Review by BZahory — DISMISSED** [review]

**Review by BZahory — APPROVED** [review]

**Review by micaelae — APPROVED** [review]

**Pull request merged**